src/metabase/models/card.clj

```diff
@@ -594,29 +676,37 @@ saved later when it is ready."
   the transaction yet. If you pass true here it is important to call the event after the cards are successfully
   created."
   ([card creator] (create-card! card creator false))
-  ([{:keys [dataset_query result_metadata dataset parameters parameter_mappings], :as card-data} creator delay-event?]
+  ([{:keys [dataset_query result_metadata dataset parameters parameter_mappings type] :as card-data} creator delay-event?]
    ;; `zipmap` instead of `select-keys` because we want to get `nil` values for keys that aren't present. Required by
    ;; `api/maybe-reconcile-collection-position!`
+   (assert-card-type-and-dataset card-data)
    (let [data-keys            [:dataset_query :description :display :name :visualization_settings
                                :parameters :parameter_mappings :collection_id :collection_position :cache_ttl]
-         card-data            (assoc (zipmap data-keys (map card-data data-keys))
-                                     :creator_id (:id creator)
-                                     :dataset (boolean (:dataset card-data))
-                                     :parameters (or parameters [])
-                                     :parameter_mappings (or parameter_mappings []))
+         card-data            (-> (zipmap data-keys (map card-data data-keys))
+                                  (assoc
+                                   :creator_id (:id creator)
+                                   :dataset    (boolean dataset)
+                                   :type       (or type
+                                                   ;; most tests will create a model using :dataset true, so we'll
+                                                   ;; respect that
+                                                   (when dataset "model")
+                                                   "question")
```

Comment:
This is redundant now that you've added the call to `ensure-type-and-dataset-are-consistent`. You can just add `:type` and `:dataset` to data-keys.
```suggestion
```
That's why I suggested to use `ensure-type-and-dataset-are-consistent` to avoid duplicating logic

Reply:
this is needed to be able to create a model with `:dataset true`, see https://github.com/metabase/metabase/pull/38074#discussion_r1473770401

Reply:
`ensure-type-and-dataset-are-consistent` sets `type=model` if `dataset=true`. The logic here is redundant and duplicated. I'm talking about these lines: https://github.com/metabase/metabase/blob/c17398925852b29f2eadd19888d55e3c70c3dfa5/src/metabase/models/card.clj#L378-L385